dedup.go

```diff
@@ -0,0 +1,427 @@
+	// lastWrite tracks the hash of the data paths
+	lastWrite     map[*Template][]byte
+	lastWriteLock sync.RWMutex
+
+	// updateCh is used to indicate an update watched data
+	updateCh chan struct{}
+
+	// wg is used to wait for a clean shutdown
+	wg sync.WaitGroup
+
+	stop     bool
+	stopCh   chan struct{}
+	stopLock sync.Mutex
+}
+
+// NewDedupManager creates a new Dedup manager
+func NewDedupManager(config *Config, clients *dep.ClientSet, brain *Brain, templates []*Template) (*DedupManager, error) {
+	d := &DedupManager{
+		config:    config,
+		clients:   clients,
+		brain:     brain,
+		templates: templates,
+		leader:    make(map[*Template]<-chan struct{}),
+		lastWrite: make(map[*Template][]byte),
+		updateCh:  make(chan struct{}, 1),
+		stopCh:    make(chan struct{}),
+	}
+	return d, nil
+}
+
+// Start is used to start the de-duplication manager
+func (d *DedupManager) Start() error {
+	log.Printf("[INFO] (dedup) starting de-duplication manager")
+
+	client, err := d.clients.Consul()
+	if err != nil {
+		return err
+	}
+	go d.createSession(client)
+
+	// Start to watch each template
+	for _, t := range d.templates {
+		go d.watchTemplate(client, t)
+	}
+	return nil
+}
+
+// Stop is used to stop the de-duplication manager
+func (d *DedupManager) Stop() error {
+	d.stopLock.Lock()
+	defer d.stopLock.Unlock()
+	if d.stop {
+		return nil
+	}
+
+	log.Printf("[INFO] (dedup) stopping de-duplication manager")
+	d.stop = true
+	close(d.stopCh)
+	d.wg.Wait()
+	return nil
+}
+
+// createSession is used to create and maintain a session to Consul
+func (d *DedupManager) createSession(client *consulapi.Client) {
+START:
+	log.Printf("[INFO] (dedup) attempting to create session")
+	session := client.Session()
+	sessionCh := make(chan struct{})
+	se := &consulapi.SessionEntry{
+		Name:     "Consul-Template de-duplication",
+		Behavior: "delete",
+		TTL:      "15s",
+	}
+	id, _, err := session.Create(se, nil)
+	if err != nil {
+		log.Printf("[ERR] (dedup) failed to create session: %v", err)
+		goto WAIT
+	}
+	log.Printf("[INFO] (dedup) created session %s", id)
+
+	// Attempt to lock each template
+	for _, t := range d.templates {
+		d.wg.Add(1)
+		go d.attemptLock(client, id, sessionCh, t)
+	}
+
+	// Renew our session periodically
+	if err := session.RenewPeriodic("15s", id, nil, d.stopCh); err != nil {
+		log.Printf("[ERR] (dedup) failed to renew session: %v", err)
+	}
+	close(sessionCh)
+
+WAIT:
+	select {
+	case <-time.After(sessionCreateRetry):
+		goto START
+	case <-d.stopCh:
+		return
+	}
+}
+
+// IsLeader checks if we are currently the leader instance
+func (d *DedupManager) IsLeader(tmpl *Template) bool {
+	d.leaderLock.RLock()
+	defer d.leaderLock.RUnlock()
+
+	lockCh, ok := d.leader[tmpl]
+	if !ok {
+		return false
+	}
+	select {
+	case <-lockCh:
+		return false
+	default:
+		return true
+	}
+}
+
+// UpdateDeps is used to update the values of the dependencies for a template
+func (d *DedupManager) UpdateDeps(t *Template, deps []dep.Dependency) {
```

Comment:
I think this function should return an error instead of just logging. We can catch that error in the runner.